packages/shared/scope-hoisting/src/link.js

```diff
@@ -140,233 +150,369 @@ export function link({
+      wrappedAssets.has(asset.id) ||
+      referencedAssets.has(asset)
+    ) {
+      needsExportsIdentifierCache.set(asset, true);
+      return true;
+    }
+
+    let isEntry = asset === bundle.getMainEntry();
+    if (isEntry && bundleNeedsMainExportsIdentifier) {
+      needsExportsIdentifierCache.set(asset, true);
+      return true;
+    }
+
+    let deps = bundleGraph.getIncomingDependencies(asset);
+    let usedSymbols = bundleGraph.getUsedSymbols(asset);
+    if (usedSymbols.has('*') && (!isEntry || asset.meta.isCommonJS)) {
+      needsExportsIdentifierCache.set(asset, true);
+      return true;
+    }
+
+    let res = deps.some(
+      dep =>
+        // Internalized async dependencies need the exports object for Promise.resolve($id$exports)
+        (dep.isAsync && bundle.hasDependency(dep)) ||
+        // If there's a dependency on the namespace, and the parent asset's exports object is used,
+        // we need to keep the exports object for $parcel$exportWildcard.
+        (!isEntry &&
+          dep.symbols.hasExportSymbol('*') &&
+          needsExportsIdentifierForAsset(
+            nullthrows(bundleGraph.getAssetWithDependency(dep)),
+          )) ||
+        // If the asset is CommonJS and there's an ES6 dependency on `default`, we need the
+        // exports identifier to call $parcel$interopDefault.
+        (asset.meta.isCommonJS &&
+          dep.meta.isES6Module &&
+          dep.symbols.hasExportSymbol('default')) ||
+        // If the asset is an ES6 module with a default export, and there's a CommonJS dependency
+        // on it, we need the exports identifier to call $parcel$defineInteropFlag.
+        (asset.meta.isES6Module &&
+          asset.symbols.hasExportSymbol('default') &&
+          dep.meta.isCommonJS &&
+          !dep.isAsync &&
+          dep.symbols.hasExportSymbol('*')) ||
+        // If one of the symbols imported by the dependency doesn't resolve, then we need the
+        // exports identifier to fall back to.
+        [...dep.symbols].some(
+          ([symbol]) => !resolveSymbol(asset, symbol, bundle).identifier,
+        ),
+    );
+
+    needsExportsIdentifierCache.set(asset, res);
+    return res;
   }
 
-  function maybeReplaceIdentifier(path: NodePath<Identifier>) {
-    let {name} = path.node;
+  function needsDeclaration(name: string) {
+    let exp = exports.get(name);
+    if (exp) {
+      let [asset, local] = exp;
+      if (asset === bundle.getMainEntry() && bundle.env.isLibrary) {
+        return true;
+      }
+
+      if (asset.meta.pureExports === false) {
+        return true;
+      }
+
+      let usedSymbols = bundleGraph.getUsedSymbols(asset);
+
+      // If the asset is CommonJS, and "default" was used but not defined, this
+      // will resolve to the $id$exports object, so we need to retain all symbols.
+      if (
+        asset.meta.isCommonJS &&
+        usedSymbols.has('default') &&
+        !asset.symbols.hasExportSymbol('default')
```

Comment:
Does this actually happen? I had thought used symbols are a subset of the asset symbols

Reply:
It happens when an ESM default import references a CJS module.